File: day1/day1.clj
(require '[clojure.java.io :as io])
(require '[clojure.string :as str])
(require 'clojure.main)

(defn read-lines [path]
  (with-open [r (io/reader path)]
    (doall (line-seq r))))

(defn parse-rotation [rotation-str]
    (let [
        dir (if (= "L" (subs rotation-str 0 1))  -1 1)
        num (Integer/parseInt (subs rotation-str 1))
    ]
    (* dir num)))


(defn sign [x]
    (cond 
        (neg? x) -1 
        (pos? x) 1
        :else 0
))

(defn rotate [current input]
    (mod (+ current input) 100))


(defn count-turns [current input]
    "Bruteforced solution to count the number of times the dial crosses 0"
    (let [
        s (sign input)
        rot (abs input)
    ]

    (loop [turns 0 rotation rot c current]
        (if (= 0 rotation)
            turns ;; return number of turns

            (let [
                dial (rotate c s) ;; next dial position
            ]

            (recur 
                ;; inc turns if dial is 0
                (if (= 0 dial) (inc turns) turns)
                (dec rotation)
                dial
            )
            )
        )
    )
    )
)

(defn count-dial-at-0 [data]
    (->> data
        (map parse-rotation)
        (reductions rotate 50)
        (filter #(= % 0))
        (count)
    )
)


(println "Part 1: " 
    (count-dial-at-0 (read-lines "input1.txt"))
)

(println "Part 2: "
    (let [
        data (read-lines "input1.txt")
        parsed-data (map parse-rotation data)
        rotations  (conj (vec parsed-data) 0)

        dial      ( ->> parsed-data
                    (reductions rotate 50)
                    (vec))
        ]

        (reduce + (map count-turns dial rotations))        
    )
)

File: day1/day1.py
import math

def read_input(fpath):
    lines = None
    with open(fpath) as inf:
        lines = inf.readlines()
    return lines


def parse_rotation(rotation):
    sign = 1.0
    if rotation[0] == "L":
        sign = -1.0
   
    return sign * int(rotation[1:])

def rotate(dial, rotation):
    return (dial + rotation) % 100

def sign(a):
    if (a < 0):
        return -1
    elif (a > 0):
        return 1
    else:
        return 0
   

def count_turns_bruteforce(dial, rotation):
    s = sign(rotation)
    rot = int(math.fabs(rotation))
    count = 0

    for r in range(rot):
        dial = rotate(dial, s)
        if dial == 0:
            count += 1    
    return count


def part1(data):
    dial = 50
    count = 0
    for line in data:
        rotation = parse_rotation(line)
        dial = rotate(dial, rotation)
        if (dial == 0):
            count += 1
    return count

def part2(data):
    dial = 50
    count = 0
    for line in data:
        rotation = parse_rotation(line)
        turns_b = count_turns_bruteforce(dial, rotation)                   
        dial = rotate(dial, rotation)
        count += turns_b

    return count


data = read_input("input1.txt")
print("Part 1:", part1(data))
print("Part 2:", part2(data))